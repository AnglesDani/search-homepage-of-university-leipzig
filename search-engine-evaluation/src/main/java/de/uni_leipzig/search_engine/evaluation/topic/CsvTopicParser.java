package de.uni_leipzig.search_engine.evaluation.topic;

import java.io.IOException;
import java.io.StringReader;
import java.util.List;
import java.util.stream.Collectors;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;

import lombok.experimental.UtilityClass;

@UtilityClass
public class CsvTopicParser
{
	private static final String CSV_FIELD_CATEGORY = "Kategorie";
	
	private static final String CSV_FIELD_CREATOR = "Ersteller";
	
	private static final String CSV_FIELD_EVALUATOR = "Bewerter";
	
	private static final String CSV_FIELD_DESCRIPTION = "Information Need";
	
	private static final String CSV_FIELD_NARRATIVE = "Criteria for Relevance";
	
	private static final String CSV_FIELD_QUERY = "Query";
	
	private static final String CSV_FIELD_BINARY_RELEVANCE_VECTOR = "Relevanz Vektor";
	
	private static final String CSV_FIELD_RECIPROCAL_RANK_RELEVANCE_VECTOR = "RR Vektor";
	
	public static List<Topic> parseTopicsFromCsvString(String csvString) throws IOException
	{
		CSVParser csvParser = CSVFormat.DEFAULT.withFirstRecordAsHeader()
				.parse(new StringReader(csvString));
		
		return csvParser.getRecords().stream()
				.map(CsvTopicParser::mapCsvRecordToTopicOrFailIfInvalid)
				.collect(Collectors.toList());
	}
	
	private static Topic mapCsvRecordToTopicOrFailIfInvalid(CSVRecord record)
	{
		return new Topic()
				.setCategory(record.get(CSV_FIELD_CATEGORY))
				.setCreator(record.get(CSV_FIELD_CREATOR))
				.setEvaluator(record.get(CSV_FIELD_EVALUATOR))
				.setDescription(record.get(CSV_FIELD_DESCRIPTION))
				.setNarrative(record.get(CSV_FIELD_NARRATIVE))
				.setQuery(record.get(CSV_FIELD_QUERY))
				.setRelevanceGainVector(RelevanceJudgement.parseRelevanceGainVector(
						record.get(CSV_FIELD_BINARY_RELEVANCE_VECTOR),
						record.get(CSV_FIELD_RECIPROCAL_RANK_RELEVANCE_VECTOR)));
	}
}
