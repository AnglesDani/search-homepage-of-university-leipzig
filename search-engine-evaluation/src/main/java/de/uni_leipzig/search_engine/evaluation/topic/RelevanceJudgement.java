package de.uni_leipzig.search_engine.evaluation.topic;

import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public enum RelevanceJudgement
{
	NOT_RELEVANT,
	RELEVANT,
	RELEVANT_FOR_MRR;

	public static RelevanceJudgement max(RelevanceJudgement a, RelevanceJudgement b)
	{
		return a.ordinal() >= b.ordinal() ? a : b;
	}

	static RelevanceJudgement[] parseRelevanceGainVector(String rawBinaryRelevanceVector, String rawReciprocalRankRelevanceVector)
	{
		RelevanceJudgement[] binaryRelevanceVector = parseBinaryRelevanceVectorOrFailIfInvalid(rawBinaryRelevanceVector);
		RelevanceJudgement[] reciprocalRankRelevanceVector = parseReciprocalRankRelevanceVectorOrFailIfInvalid(rawReciprocalRankRelevanceVector);
		
		return IntStream.range(0, binaryRelevanceVector.length)
			.mapToObj(i -> RelevanceJudgement.max(binaryRelevanceVector[i], reciprocalRankRelevanceVector[i]))
			.collect(Collectors.toList())
			.toArray(new RelevanceJudgement[binaryRelevanceVector.length]);
	}
	
	private static RelevanceJudgement[] parseRelevanceJudgementsFromDigits(String number)
	{
		return number.chars()
				.mapToObj(character -> RelevanceJudgement.values()[Character.getNumericValue((char) character)])
				.collect(Collectors.toList())
				.toArray(new RelevanceJudgement[number.length()]);
	}
	
	private static RelevanceJudgement[] parseBinaryRelevanceVectorOrFailIfInvalid(String rawBinaryRelevanceVector)
	{
		RelevanceJudgement[] ret = RelevanceJudgement.parseRelevanceJudgementsFromDigits(rawBinaryRelevanceVector);
		
		if(ret.length != rawBinaryRelevanceVector.length() || 
				Arrays.stream(ret).anyMatch(jugement -> jugement.equals(RelevanceJudgement.RELEVANT_FOR_MRR)))
		{
			throw new RuntimeException("Couldnt parse the relevance vector '"+ rawBinaryRelevanceVector 
					+"'. Only digits 0 and 1 are allowed.");
		}

		return ret;
	}
	
	private static RelevanceJudgement[] parseReciprocalRankRelevanceVectorOrFailIfInvalid(String rawReciprocalRankRelevanceVector)
	{
		RelevanceJudgement[] ret = RelevanceJudgement.parseRelevanceJudgementsFromDigits(rawReciprocalRankRelevanceVector);
		
		if(ret.length != rawReciprocalRankRelevanceVector.length() ||
				Arrays.stream(ret).anyMatch(jugement -> jugement.equals(RelevanceJudgement.RELEVANT)) ||
				Arrays.stream(ret).filter(jugement -> jugement.equals(RelevanceJudgement.RELEVANT_FOR_MRR)).count() > 1)
		{
			throw new RuntimeException("Couldnt parse the reciprocal rank relevance vector '"+ 
					rawReciprocalRankRelevanceVector +"'. Only digits 0 and one single 2 are allowed.");
		}
			
		return ret;
	}
}
