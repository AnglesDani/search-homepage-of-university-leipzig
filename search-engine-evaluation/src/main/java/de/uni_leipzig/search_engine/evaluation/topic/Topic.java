package de.uni_leipzig.search_engine.evaluation.topic;

import lombok.Data;
import lombok.experimental.Accessors;

@Data
@Accessors(chain=true)
public class Topic
{
	private String query;
	
	private String description;
	
	private String narrative;
	
	private String category;
	
	private String creator;
	
	private String evaluator;
	
	private RelevanceJudgement[] relevanceGainVector;
}
