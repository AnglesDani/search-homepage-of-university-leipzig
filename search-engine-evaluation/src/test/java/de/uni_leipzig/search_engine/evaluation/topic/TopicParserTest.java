package de.uni_leipzig.search_engine.evaluation.topic;

import java.nio.file.Files;
import java.nio.file.Paths;

import org.approvaltests.Approvals;
import org.junit.Test;

import de.uni_leipzig.search_engine.evaluation.topic.CsvTopicParser;
import lombok.SneakyThrows;

public class TopicParserTest
{
	@Test
	@SneakyThrows
	public void checkThatTopicWithOneLineCouldBeParsed()
	{
		String csvString = readFile("src/test/resources/one_line_search_evaluation.csv");
		
		Approvals.verify(CsvTopicParser.parseTopicsFromCsvString(csvString));
	}
	
	@Test
	@SneakyThrows
	public void checkThatTopicWithMultipleLinesCouldBeParsed()
	{
		String csvString = readFile("src/test/resources/multi_line_search_evaluation.csv");
		
		Approvals.verify(CsvTopicParser.parseTopicsFromCsvString(csvString));
	}
	
	@Test(expected=RuntimeException.class)
	@SneakyThrows
	public void checkThatTopicWithInvalidReciprocalRankJudgementCouldNotBeParsed()
	{
		String csvString = readFile("src/test/resources/one_line_search_evaluation_with_invalid_rr_range.csv");
		
		CsvTopicParser.parseTopicsFromCsvString(csvString);
	}
	
	@Test(expected=RuntimeException.class)
	@SneakyThrows
	public void checkThatTopicWithMultipleRelevantReciprocalRankJudgementCouldNotBeParsed()
	{
		String csvString = readFile("src/test/resources/one_line_search_evaluation_with_multiple_valid_rr_entries.csv");
		
		CsvTopicParser.parseTopicsFromCsvString(csvString);
	}
	
	@Test(expected=RuntimeException.class)
	@SneakyThrows
	public void checkThatTopicWithInvalidRelevanceJudgmentCouldNotBeParsed()
	{
		String csvString = readFile("src/test/resources/one_line_search_evaluation_with_invalid_relevance_vector.csv");
		
		CsvTopicParser.parseTopicsFromCsvString(csvString);
	}
	
	@SneakyThrows
	private static final String readFile(String path)
	{
		return new String(Files.readAllBytes(Paths.get(path)));
	}
}
