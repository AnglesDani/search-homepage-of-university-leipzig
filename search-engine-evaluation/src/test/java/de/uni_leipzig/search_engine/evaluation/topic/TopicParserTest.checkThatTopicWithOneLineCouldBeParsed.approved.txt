[Topic(query=droste email, description=E-Mail-Adresse von Professor Droste erfahren, narrative=Gesucht ist eine Webseite, die von dem Mitarbeiter Professor Droste der Universität Leipzig die E-Mail enthält. Es fallen also Publikationen weg, bei denen im Header die E-Mail von Professor Droste enthalten ist da er Author der Publikation war., category=Personensuche/Kontaktdaten, creator=Maik, evaluator=Maik, relevanceGainVector=[NOT_RELEVANT, RELEVANT_FOR_MRR, NOT_RELEVANT, NOT_RELEVANT, NOT_RELEVANT, NOT_RELEVANT, NOT_RELEVANT, NOT_RELEVANT, NOT_RELEVANT, NOT_RELEVANT])]